spark3/src/main/java/org/apache/iceberg/actions/Spark3SnapshotAction.java

```diff
@@ -0,0 +1,148 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.actions;
+
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.iceberg.Snapshot;
+import org.apache.iceberg.SnapshotSummary;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.TableProperties;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.spark.SparkTableUtil;
+import org.apache.iceberg.spark.source.SparkTable;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.catalyst.TableIdentifier;
+import org.apache.spark.sql.catalyst.analysis.NoSuchNamespaceException;
+import org.apache.spark.sql.catalyst.analysis.TableAlreadyExistsException;
+import org.apache.spark.sql.connector.catalog.CatalogPlugin;
+import org.apache.spark.sql.connector.catalog.Identifier;
+import org.apache.spark.sql.connector.catalog.StagedTable;
+import org.apache.spark.sql.connector.catalog.TableCatalog;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.collection.JavaConverters;
+
+/**
+ * Creates a new Iceberg table based on a source Spark table. The new Iceberg table will
+ * have a different data and metadata directory allowing it to exist independently of the
+ * source table.
+ */
+class Spark3SnapshotAction extends Spark3CreateAction implements SnapshotAction {
+  private static final Logger LOG = LoggerFactory.getLogger(Spark3SnapshotAction.class);
+
+  private String destTableLocation = null;
+
+  Spark3SnapshotAction(SparkSession spark, CatalogPlugin sourceCatalog,
+                       Identifier sourceTableName, CatalogPlugin destCatalog,
+                       Identifier destTableName) {
+    super(spark, sourceCatalog, sourceTableName, destCatalog, destTableName);
+  }
+
+  @Override
+  public Long execute() {
+    StagedTable stagedTable;
+    Table icebergTable;
+
+    try {
+      stagedTable = destCatalog().stageCreate(destTableName(), v1SourceTable().schema(),
+          sourcePartitionSpec(), buildPropertyMap());
+      icebergTable = ((SparkTable) stagedTable).table();
+
+      if (!icebergTable.properties().containsKey(TableProperties.DEFAULT_NAME_MAPPING)) {
+        assignDefaultTableNameMapping(icebergTable);
+      }
+    } catch (TableAlreadyExistsException taeException) {
+      throw new IllegalArgumentException("Cannot create snapshot because a table already exists with that name",
+          taeException);
+    } catch (NoSuchNamespaceException nsnException) {
+      throw new IllegalArgumentException("Cannot create snapshot because the namespace given does not exist",
+          nsnException);
+    }
+
+    boolean threw = true;
+    try {
+      String stagingLocation = icebergTable.location() + "/" + ICEBERG_METADATA_FOLDER;
```

Comment:
I don't think this is correct. I think we should allow custom data and metadata folders in SNAPSHOT.

Here is how get the metadata location internally:

```
  def getMetadataLocation(table: Table): String = {
    table.properties.getOrDefault(
      TableProperties.WRITE_METADATA_LOCATION,
      table.location + "/" + AdtConstants.ICEBERG_METADATA_FOLDER)
  }
```

I believe we should do the same here.

Reply:
I don't really have an opinion on whether we need to customize this in snapshot tables. I think for migration it is a good idea. For snapshots, I'd probably opt to keep it simple.

Reply:
I think the simplest option for us is not to do anything for SNAPSHOT and let users define all properties including data and metadata location. That's why we need the logic in `getMetadataLocation` to pick the correct metadata location for staging.